cuML/src/random_projection/utils.hxx

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "rproj_c.h"
+#include <random/rng.h>
+
+using namespace MLCommon::Random;
+
+inline void sample_without_replacement(size_t n_population, size_t n_samples,
+											int* indices, size_t& indices_idx)
+{
+    std::random_device dev;
+    std::mt19937 gen(dev());
+    std::uniform_int_distribution<int> uni_dist(0, n_population-1);
+
+    std::unordered_set<int> s;
+
+    for (size_t i = 0; i < n_samples; i++)
+    {
+        int rand_idx = uni_dist(gen);
+        while (s.find(rand_idx) != s.end())
+        {
+            rand_idx = uni_dist(gen);
+        }
+        s.insert(rand_idx);
+        indices[indices_idx] = rand_idx;
+        indices_idx++;
+    }
+}
+
+inline size_t binomial(size_t n, double p)
+{
+    std::random_device dev;
+    std::mt19937 gen(dev());
+    std::binomial_distribution<size_t> bin_dist(n, p);
+    return bin_dist(gen);
+}
+
+inline double check_density(double density, size_t n_features)
+{
+    if (density == -1.0)
+    {
+        return 1.0 / sqrt(n_features);
+    }
+    return density;
+}
+
+/**
+ * @brief computes minimum target dimension to preserve information according to error tolerance (eps parameter)
+ * @input param n_samples: number of samples
+ * @input param eps: error tolerance
+ * @return minimum target dimension
+ */
+inline size_t johnson_lindenstrauss_min_dim(size_t n_samples, double eps)
+{
+    ASSERT(eps > 0.0 && eps < 1.0,
+                "Parameter eps: must be in range ]0, 1[");
```

Comment:
Nitpick: isn't the standard notation for the open set `(0,1)`?

Reply:
^^ Yes, I think different countries have different notations for intervals. I took the one from sklearn (this model was made by a French author). But yes, I can use this notation instead.